upgrade isort to v5.5.2

vporta opened the pull request:
<!-- Thank you for sending a PR and taking the time to improve Hydra -->

## Motivation

Fix issue #751 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)
Run tests as per the contributing documentation.
```
$ nox -s lint
$ FIX=1 nox -s lint_plugins
$ pytest
$ pre-commit run --all-files  # to check the new isort hooks pass
```


## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)
issue #751

omry commented on .isort.cfg:
```diff
@@ -1,15 +1,24 @@
+known_first_party=hydra,hydra_app,configen,example
 skip=
     plugins
+    ,__init__.py
```
Why do you need to skip `__init__.py`?

vporta replied:
No problem! isort doesn't default skip `__init__.py` anymore: https://pycqa.github.io/isort/docs/upgrade_guides/5.0.0/#not_skip

omry replied:
I see. as a followup, can you experiment with not excluding __init__.py from isort and see if we can make that work?

vporta replied:
Sure.

vporta replied:
Removed skip of **init**.py from the config and skip list and tests pass.

omry commented on .isort.cfg:
```diff
@@ -1,15 +1,24 @@
     ,hydra/grammar/gen
     ,tools/configen/example/gen
-    ,tools/configen/tests/test_modules/expected
+    ,tools/configen/tests/test_modules/expected
```
can you add a trailing newline?

vporta replied:
I'll add it for this and in the `noxfile.py`

vporta replied:
Does this work? Trailing newlines:
![Screen Shot 2020-09-17 at 12 06 40 PM](https://user-images.githubusercontent.com/14094630/93497553-8968db00-f8de-11ea-8b61-ad41be517152.png)
![Screen Shot 2020-09-17 at 12 07 23 PM](https://user-images.githubusercontent.com/14094630/93497555-8b329e80-f8de-11ea-8fe1-5fc5657233fb.png)

omry replied:
yes

omry commented on noxfile.py:
```diff
@@ -212,6 +212,7 @@ def lint(session):
         ".git",
         "website",
         "plugins",
+        "__init__.py",
```
ditto

vporta replied:
I think I'd have to exclude it `__init__.py` hear as well

omry reviewed (changes requested):
A few minor comments/questions.
good job, thanks!

omry commented:
How are you running isort now for the whole project?
we still have to run isort separately for each plugin, tool and examples dir?

vporta commented:
> How are you running isort now for the whole project?
In root dir, I first run `$ isort . --check --diff` to check only for differences w/out modifying the files. Then `isort .` 
I do the same in plugins, tool, examples. 
> we still have to run isort separately for each plugin, tool and examples dir?
Yes I believe so. When I run `$ isort . --check --diff` in plugins/ there is no output. This is helpful for me to see which config is being used `$ isort --show-config`. When it's run in plugins/ dir., it correctly finds the isort config in plugins/. Doing the same command in tool or examples reads the isort config in the root dir.

omry commented:
Currently plugins are excluded in the top level config to prevent isort 4 to run on them with the wrong config.

vporta commented:
> Currently plugins are excluded in the top level config to prevent isort 4 to run on them with the wrong config.

Even when removing the plugins skip in the top level config file, isort reads the correct config file in plugins/ and will sort according to this config file. I confirmed this by running `$ isort --show-config` in plugins/.

omry commented:
> > Currently plugins are excluded in the top level config to prevent isort 4 to run on them with the wrong config.
> 
> Even when removing the plugins skip in the top level config file, isort reads the correct config file in plugins/ and will sort according to this config file. I confirmed this by running `$ isort --show-config` in plugins/ (with the plugins dir no longer skipped in the top level config).

If you are running isort at the top level of Hydra, does it now correctly sort plugin, examples and tools imports?
If so, we can simplify the noxfile to call isort once for the whole project.

vporta commented:
> > > Currently plugins are excluded in the top level config to prevent isort 4 to run on them with the wrong config.
> > 
> > 
> > Even when removing the plugins skip in the top level config file, isort reads the correct config file in plugins/ and will sort according to this config file. I confirmed this by running `$ isort --show-config` in plugins/ (with the plugins dir no longer skipped in the top level config).
> 
> If you are running isort at the top level of Hydra, does it now correctly sort plugin, examples and tools imports?
> If so, we can simplify the noxfile to call isort once for the whole project.

I tried this initially, but one isort command won't correctly sort plugins/. Specifically `hydra_plugins` is firstparty in plugins/, but thirdparty not in plugins/. The isort author has a workaround using the src_paths option, but didn't work for me.

omry commented:
If you share more details about what you tried with src_paths and how it failed maybe @timothycrosley will have another idea.

vporta commented:
> If you share more details about what you tried with src_paths and how it failed maybe @timothycrosley will have another idea.

- I removed the config in plugins/ directory. 
- In the top level config file: 
     - set to nothing: `src_paths=`
     - removed plugins from `skip=`
     - add hydra and hydra_plugins to `known_first_party=`

The final top level config: 
```
[settings]
multi_line_output=3
include_trailing_comma=true
force_grid_wrap=0
use_parentheses=True
line_length=88
ensure_newline_before_comments=true
known_libs=pkg_resources,setuptools
sections=
    FUTURE
    STDLIB
    LIBS
    THIRDPARTY
    FIRSTPARTY
    LOCALFOLDER
default_section=THIRDPARTY
known_first_party=hydra,hydra_app,configen,example,hydra_plugins
src_paths=
skip=
    ,.nox
    ,hydra/grammar/gen
    ,tools/configen/example/gen
    ,tools/configen/tests/test_modules/expected
```

Running `isort . --check --diff`; here is one (out of around 25) examples of an incorrect import sort error I received now in the examples directory: 
```
ERROR: /hydra/examples/plugins/example_sweeper_plugin/tests/test_example_sweeper_plugin.py Imports are incorrectly sorted and/or formatted.
--- /hydra/examples/plugins/example_sweeper_plugin/tests/test_example_sweeper_plugin.py:before    2020-09-16 14:56:50
+++ /hydra/examples/plugins/example_sweeper_plugin/tests/test_example_sweeper_plugin.py:after     2020-09-17 15:24:41.030869
@@ -1,5 +1,6 @@
 # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
 import pytest
+
 from hydra.core.plugins import Plugins
 from hydra.plugins.sweeper import Sweeper
 from hydra.test_utils.launcher_common_tests import (
@@ -8,7 +9,6 @@
     LauncherTestSuite,
 )
 from hydra.test_utils.test_utils import TSweepRunner
-
 from hydra_plugins.example_sweeper_plugin.example_sweeper import ExampleSweeper
```

Another in the plugins directory: 
```
ERROR: /Users/vincentporta/Desktop/OpenSource/hydra/plugins/hydra_ax_sweeper/tests/test_ax_sweeper_plugin.py Imports are incorrectly sorted and/or formatted.
--- /Users/vincentporta/Desktop/OpenSource/hydra/plugins/hydra_ax_sweeper/tests/test_ax_sweeper_plugin.py:before        2020-09-16 14:56:50
+++ /Users/vincentporta/Desktop/OpenSource/hydra/plugins/hydra_ax_sweeper/tests/test_ax_sweeper_plugin.py:after 2020-09-17 15:24:41.840108
@@ -7,12 +7,12 @@
 from typing import Any, List
 
 import pytest
+from omegaconf import DictConfig, OmegaConf
+
 from hydra.core.hydra_config import HydraConfig
 from hydra.core.plugins import Plugins
 from hydra.plugins.sweeper import Sweeper
 from hydra.test_utils.test_utils import TSweepRunner, chdir_plugin_root
-from omegaconf import DictConfig, OmegaConf
-
 from hydra_plugins.hydra_ax_sweeper.ax_sweeper import AxSweeper  # type: ignore
```

omry commented:
> add hydra and hydra_plugins to known_first_party=
Here is a problem:
for plugins, Hydra is third party.
for hydra plugins are third party.

You do need two different configs.
The question is if isort can be made to have effective different configs per directory based on config files along the parents of that dir.

vporta commented:
Yes two different configs.  

Kind regards,
Vincent


> On Sep 17, 2020, at 3:49 PM, Omry Yadan <notifications@github.com> wrote:
> 
> ﻿
> add hydra and hydra_plugins to known_first_party=
> 
> Here is a problem:
> for plugins, Hydra is third party.
> for hydra plugins are third party.
> 
> You do need two different configs.
> The question is if isort can be made to have effective different configs per directory based on config files along the parents of that dir.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

vporta commented:
I’m unsure if isort can do what you’re asking. 


> On Sep 17, 2020, at 6:44 PM, Vincent Porta <vporta7@gmail.com> wrote:
> 
> ﻿Yes two different configs.  
> 
> Kind regards,
> Vincent
> 
> 
>>> On Sep 17, 2020, at 3:49 PM, Omry Yadan <notifications@github.com> wrote:
>>> 
>> ﻿
>> add hydra and hydra_plugins to known_first_party=
>> 
>> Here is a problem:
>> for plugins, Hydra is third party.
>> for hydra plugins are third party.
>> 
>> You do need two different configs.
>> The question is if isort can be made to have effective different configs per directory based on config files along the parents of that dir.
>> 
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub, or unsubscribe.

timothycrosley commented:
:wave: Hi! I think my documentation was a bit misleading, I apologize for that, and will do the best to improve it!
What was meant was that isort 5 only allows one config per an execution. So if you have a set-up like 

```
~/.isort.cfg
company/.isort.cfg
company/project/isort.cfg
```

Where you expect your personal and company configs to be merged, you need to change that expectation and instead make your full config live within project:

~/.isort.cfg~
~company/.isort.cfg~
```
company/project/isort.cfg
```

However, if you have a setup, like hydra does, such as:

```
project/.isort.cfg
project/main/.isort.cfg
project/plugins/.isort.cfg
```

You will want to keep main and plugins with different .isort.cfg likely, unless the module placement logic is intended to stay the same. Importantly, then you will need a separate execution of isort per a config:

```
isort main
isort plugins
```

I'd imagine then for precommit, you would need 2 executions as well, taking advantadge of it's built-in filtering to make sure each only targets the appropriate set of files: https://pre-commit.com/#regular-expressions.

Hopefully this is helpful, I'm open to suggestions on both how to improve the documentation and the support for multi package repositories.

Thanks!

~Timothy

~Timothy

omry commented:
Hi @timothycrosley!
Thanks for the explanation, this kind of functionality is workable, at least when using nox. I am not exactly sure how to get it to work to the right thing with pre-commit.

Regardless, the ideal setup is such that I run isort once in the root of the project and everything "just works (tm)".
For my use case, this means that for each directory there is a single effective config, which is the first isort config encountered along the parent path (I think the "search" should be terminated at cwd).
so:

```
hydra/.isort.cfg # config for the top level project
hydra/hydra/main.py       # effective config is `hydra/.isort.cfg`
hydra/plugins/.isort.cfg    # config for the plugins directory and sub directories
      plugins/plugin1/foo.py  # effective config is `hydra/plugins/.isort.cfg`
      plugins/plugin2/bar.py  # effective config is `hydra/plugins/.isort.cfg`
```
Based on what I am seeing in the pr from @vporta (Thanks!!), isort is now powerful enough to deal with all my plugins with single config.
Should a specific plugin require a unique config, I can just drop a complete `.isort.cfg` in that plugin's dir.

In this scenario, there are no config merges (which - if I understand correctly - is something you want to avoid). Instead a single config file applies to all the sub directories under it. Furthermore the effective config is re-evaluated when entering a new directory.

I hope the explanation of the perfect solution for me is clear.
Thanks for working on isort!

timothycrosley commented:
I'll be thinking on what the cleanest way to support this in a single execution on the isort project, and have opened an issue on the isort project to do so.

In the meantime, for pre-commit, I believe, you simply would need to list out each execution of isort separately per a config used:

```
  - repo: https://github.com/pycqa/isort
    rev: 5.5.2
    hooks:
      - id: isort
        files: 'hydra/.*'
      - id: isort
        files: 'plugins/.*'
```

omry commented:
Great, thanks @timothycrosley!
@vporta , mind updating the PR pre-commit config as suggested by @timothycrosley ?

Also, can you add another config (and run in nox and pre-commit) for the directories under tools? (those are standalone projects).

vporta commented:
> Great, thanks @timothycrosley!
> @vporta , mind updating the PR pre-commit config as suggested by @timothycrosley ?

Will do! 
> 
> Also, can you add another config (and run in nox and pre-commit) for the directories under tools? (those are standalone projects).

Sure, wouldn't I need to update the pre-commit config with (below) as well? 

```
- id: isort
  files: 'tools/.*'
```

omry commented:
yup.

vporta commented:
The Test Plan tests pass. Here are the changes: 

In the pre-commit file, I added the new hook for plugins and tools.

I dropped a new isort config in `tools/`:
```
[settings]
multi_line_output=3
include_trailing_comma=true
force_grid_wrap=0
use_parentheses=true
line_length=88
ensure_newline_before_comments=true
float_to_top=true
sections=
    FUTURE
    STDLIB
    THIRDPARTY
    FIRSTPARTY
    LOCALFOLDER
default_section=THIRDPARTY
known_first_party=tools
```

In the isort config file in 'tools/', I set tools to `known_first_party`. 

Since a new isort config file is dropped in 'tools/', I think it should be skipped at the top level (like plugins), so in the top-level isort config, I added 'tools/' to `skip=`. By the way, we might be able to remove `tools/configen/example/gen
    ,tools/configen/tests/test_modules/expected` from `skip` now (as well as in the `skiplist` in noxfile). The top level isort config is now:  
```
[settings]
multi_line_output=3
include_trailing_comma=true
force_grid_wrap=0
use_parentheses=True
line_length=88
ensure_newline_before_comments=true
known_libs=pkg_resources,setuptools
sections=
    FUTURE
    STDLIB
    LIBS
    THIRDPARTY
    FIRSTPARTY
    LOCALFOLDER
default_section=THIRDPARTY
known_first_party=hydra,hydra_app
skip=
    plugins
    ,tools
    ,.nox
    ,hydra/grammar/gen
    ,tools/configen/example/gen
    ,tools/configen/tests/test_modules/expected
```

To continue consistency I skipped 'tools' in the `skiplist` in `noxfile.py` as well:  
Before:
https://github.com/facebookresearch/hydra/blob/5a34a01eaa0f0426d967e918a3ecd8ac6fcf9f47/noxfile.py#L216-L224
After:
```
skiplist = apps + [
        ".git",
        "website",
        "plugins",
        "tools",
        ".nox",
        "hydra/grammar/gen",
        "tools/configen/example/gen",
        "tools/configen/tests/test_modules/expected",
    ]
```

Am I missing anything?

omry commented:
tools is not a Python module, I don't think you should add it to known first party.
The module for `tools/configen/example/gen` is `example/gen`.

omry commented:
Is configen detected as first party in tools/configen?

vporta commented:
> tools is not a Python module, I don't think you should add it to known first party.
> The module for `tools/configen/example/gen` is `example/gen`.

I don't see an examples directory. I see `tools/configen/example/config`

vporta commented:
> Is configen detected as first party in tools/configen?

Currently, tools is considered first party, but configen is in the same import placement as master. For example, in `tools/configen/example/my_app.py`:
```
import hydra

# Generated config dataclasses
from example.config.configen.samples.my_module import AdminConf, UserConf
from hydra.core.config_store import ConfigStore

# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
from omegaconf import DictConfig

# Underlying objects
from configen.samples.my_module import Admin, User
```

omry commented:
> > Is configen detected as first party in tools/configen?
> 
> Currently, tools/configen is considered first party, but configen is in the same import placement as master. For example, in `tools/configen/example/my_app.py`:
> 
> ```
> import hydra
> 
> # Generated config dataclasses
> from example.config.configen.samples.my_module import AdminConf, UserConf
> from hydra.core.config_store import ConfigStore
> 
> # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
> from omegaconf import DictConfig
> 
> # Underlying objects
> from configen.samples.my_module import Admin, User
> ```

This seems okay, except for the Copyright comment that moved for some reason.

omry commented:
> > Is configen detected as first party in tools/configen?
> 
> Currently, tools/configen is considered first party, but configen is in the same import placement as master. For example, in `tools/configen/example/my_app.py`:
> 
> ```
> import hydra
> 
> # Generated config dataclasses
> from example.config.configen.samples.my_module import AdminConf, UserConf
> from hydra.core.config_store import ConfigStore
> 
> # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
> from omegaconf import DictConfig
> 
> # Underlying objects
> from configen.samples.my_module import Admin, User
> ```

I am confused:
The change to the sort order (that moved the copyright header) is not in the diff.
where did it came from? how come the tests are passing if this is not in the diff?

vporta commented:
> > > Is configen detected as first party in tools/configen?
> > 
> > 
> > Currently, tools/configen is considered first party, but configen is in the same import placement as master. For example, in `tools/configen/example/my_app.py`:
> > ```
> > import hydra
> > 
> > # Generated config dataclasses
> > from example.config.configen.samples.my_module import AdminConf, UserConf
> > from hydra.core.config_store import ConfigStore
> > 
> > # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
> > from omegaconf import DictConfig
> > 
> > # Underlying objects
> > from configen.samples.my_module import Admin, User
> > ```
> 
> I am confused:
> The change to the sort order (that moved the copyright header) is not in the diff.
> where did it came from? how come the tests are passing if this is not in the diff?

It came from my local branch and I haven't pushed up the code yet, will do it soon.

vporta commented:
> > > Is configen detected as first party in tools/configen?
> > 
> > 
> > Currently, tools/configen is considered first party, but configen is in the same import placement as master. For example, in `tools/configen/example/my_app.py`:
> > ```
> > import hydra
> > 
> > # Generated config dataclasses
> > from example.config.configen.samples.my_module import AdminConf, UserConf
> > from hydra.core.config_store import ConfigStore
> > 
> > # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
> > from omegaconf import DictConfig
> > 
> > # Underlying objects
> > from configen.samples.my_module import Admin, User
> > ```
> 
> This seems okay, except for the Copyright comment that moved for some reason.

 For some reason isort moved the Copyright comment (above omegaconf) when omegaconf was sorted/moved. This is the updated version after running isort in 'tools`:
```
# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
import hydra

# Generated config dataclasses
from example.config.configen.samples.my_module import AdminConf, UserConf
from hydra.core.config_store import ConfigStore
from omegaconf import DictConfig

# Underlying objects
from configen.samples.my_module import Admin, User
```

It looks correct. configen is first party and the rest are third party.

omry commented:
My guess is that isort associate the the copyright header with the first import.
Can you test that by causing a sort to happen and see if that the copyright header is moved?

if that's the case, I think inserting a space after the copyright comment would fix it. do you know if isort can do it?

vporta commented:
> My guess is that isort associate the the copyright header with the first import.
> Can you test that by causing a sort to happen and see if that the copyright header is moved?
> 
> if that's the case, I think inserting a space after the copyright comment would fix it. do you know if isort can do it?

Just tested it. I moved the first import and the copyright comment `import hydra` here:
```



# Generated config dataclasses
from example.config.configen.samples.my_module import AdminConf, UserConf
from hydra.core.config_store import ConfigStore
from omegaconf import DictConfig
# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
import hydra
# Underlying objects
from configen.samples.my_module import Admin, User
```

Then ran isort and it moved it back: 
```
# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
import hydra

# Generated config dataclasses
from example.config.configen.samples.my_module import AdminConf, UserConf
from hydra.core.config_store import ConfigStore
from omegaconf import DictConfig

# Underlying objects
from configen.samples.my_module import Admin, User
```

vporta commented:
> > My guess is that isort associate the the copyright header with the first import.
> > Can you test that by causing a sort to happen and see if that the copyright header is moved?
> > if that's the case, I think inserting a space after the copyright comment would fix it. do you know if isort can do it?
> 
> Just tested it. I moved the first import and the copyright comment `import hydra` here:
> 
> ```
> 
> 
> 
> # Generated config dataclasses
> from example.config.configen.samples.my_module import AdminConf, UserConf
> from hydra.core.config_store import ConfigStore
> from omegaconf import DictConfig
> # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
> import hydra
> # Underlying objects
> from configen.samples.my_module import Admin, User
> ```
> 
> Then ran isort and it moved it back:
> 
> ```
> # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
> import hydra
> 
> # Generated config dataclasses
> from example.config.configen.samples.my_module import AdminConf, UserConf
> from hydra.core.config_store import ConfigStore
> from omegaconf import DictConfig
> 
> # Underlying objects
> from configen.samples.my_module import Admin, User
> ```

When removing the other comments and running isort it groups the imports like this:
```
# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
import hydra
from example.config.configen.samples.my_module import AdminConf, UserConf
from hydra.core.config_store import ConfigStore
from omegaconf import DictConfig

from configen.samples.my_module import Admin, User
```
Yep you're right, adding a space under a comment doesn't move the import with the comment

omry commented:
Your diff now contains unrelated changes. can you clean it up?
It should only contains changes you are doing. (rebase, not merge).

vporta commented:
> Your diff now contains unrelated changes. can you clean it up?
> It should only contains changes you are doing. (rebase, not merge).

I moved the changes over to another branch and will open a new PR. Is that alright?

omry commented:
It's fine, you can also try to clean things up and force push.
your call. If you open a new branch be sure to close this one.

vporta commented:
> It's fine, you can also try to clean things up and force push.
> your call. If you open a new PR be sure to close this one.

Thanks I'll close this then.
